### PR TITLE
[Build][Test] TF 2.16.1 Inference SM/EC2 Image release

### PR DESCRIPTION
**Pull request by shantanutrip**
*GitHub Issue #, if available:*

**Note**: 
- If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

- All PR's are checked weekly for staleness. This PR will be closed if not updated in 30 days.

### Description
[Build][Test] TF 2.16.1 Inference SM/EC2 Image release

### Tests run
<details>
<summary>Confused on how to run tests? Try using the helper utility...</summary>

Assuming your remote is called `origin` (you can find out more with `git remote -v`)...
  
- Run default builds and tests for a particular buildspec - also commits and pushes changes to remote; Example:

`python src/prepare_dlc_dev_environment.py -b </path/to/buildspec.yml> -cp origin`

- Enable specific tests for a buildspec or set of buildspecs - also commits and pushes changes to remote; Example:

`python src/prepare_dlc_dev_environment.py -b </path/to/buildspec.yml> -t sanity_tests -cp origin`

- Restore TOML file when ready to merge

`python src/prepare_dlc_dev_environment.py -rcp origin`
</details>

**NOTE: By default, docker builds are disabled. In order to build your container, please update dlc_developer_config.toml and specify the framework to build in "build_frameworks"**
- [ ] I have run builds/tests on commit <INSERT COMMIT ID> for my changes.

**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**
<details>
<summary>Expand</summary>

- [ ] `sagemaker_remote_tests = true`
- [ ] `sagemaker_efa_tests = true`
- [ ] `sagemaker_rc_tests = true`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

</details>

### Formatting
- [ ] I have run `black -l 100` on my code (formatting tool: https://black.readthedocs.io/en/stable/getting_started.html)

#### Builds to Execute
<details>
<summary>Expand</summary>

Fill out the template and click the checkbox of the builds you'd like to execute

*Note: Replace with <X.Y> with the major.minor framework version (i.e. 2.2) you would like to start.*

- [ ] build_pytorch_training_<X.Y>_sm
- [ ] build_pytorch_training_<X.Y>_ec2

- [ ] build_pytorch_inference_<X.Y>_sm
- [ ] build_pytorch_inference_<X.Y>_ec2
- [ ] build_pytorch_inference_<X.Y>_graviton

- [ ] build_tensorflow_training_<X.Y>_sm
- [ ] build_tensorflow_training_<X.Y>_ec2

- [x] build_tensorflow_inference_2.16_sm
- [x] build_tensorflow_inference_2.16_ec2
- [ ] build_tensorflow_inference_<X.Y>_graviton
</details>

### PR Checklist 
<details>
<summary>Expand</summary>

- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

#### NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ec2_benchmark_tests = true` or `sagemaker_benchmark_tests = true`
</details>

### Pytest Marker Checklist
<details>
<summary>Expand</summary>

- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type
</details>

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by arjkesh** [comment]
you can add the images to release_images_patches.yml in this PR as well

File: tensorflow/inference/docker/2.16/py3/Dockerfile.cpu [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+
+FROM base_image AS ec2
+
+LABEL maintainer="Amazon AI"
+LABEL dlc_major_version="1"
+
+ARG PYTHON=python3.10
+ARG PYTHON_PIP=python3-pip
+ARG PIP=pip3
+ARG PYTHON_VERSION=3.10.13
```
Comment (junpuf) [inline tensorflow/inference/docker/2.16/py3/Dockerfile.cpu:35]: can probably go to [3.10.14](https://www.python.org/ftp/python/3.10.14/)

File: tensorflow/inference/docker/2.16/py3/cu122/Dockerfile.gpu [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+
+FROM base_image AS ec2
+
+LABEL maintainer="Amazon AI"
+LABEL dlc_major_version="1"
+
+ARG PYTHON=python3.10
+ARG PYTHON_PIP=python3-pip
+ARG PIP=pip3
+ARG PYTHON_VERSION=3.10.13
```
Comment (junpuf) [inline tensorflow/inference/docker/2.16/py3/cu122/Dockerfile.gpu:35]: same as above, we can go to 3.10.14

File: tensorflow/inference/docker/2.16/py3/cu122/Dockerfile.gpu [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+
+FROM ec2 AS sagemaker
+
+LABEL maintainer="Amazon AI"
+LABEL dlc_major_version="1"
+
+# Specify accept-bind-to-port LABEL for inference pipelines to use SAGEMAKER_BIND_TO_PORT
+# https://docs.aws.amazon.com/sagemaker/latest/dg/inference-pipeline-real-time.html
+LABEL com.amazonaws.sagemaker.capabilities.accept-bind-to-port=true
+
```
Comment (junpuf) [inline tensorflow/inference/docker/2.16/py3/cu122/Dockerfile.gpu:219]: Add a LABEL like [this](https://github.com/aws/deep-learning-containers/blob/master/pytorch/inference/docker/2.3/py3/cu121/Dockerfile.gpu#L304)

**Review by junpuf — CHANGES_REQUESTED** [review]

File: tensorflow/inference/docker/2.16/py3/cu122/Dockerfile.gpu (hunk 3 above)
Comment (shantanutrip) [reply tensorflow/inference/docker/2.16/py3/cu122/Dockerfile.gpu:220]: Thanks for point this! Added

File: tensorflow/inference/docker/2.16/py3/cu122/Dockerfile.gpu (hunk 2 above)
Comment (shantanutrip) [reply tensorflow/inference/docker/2.16/py3/cu122/Dockerfile.gpu]: Thanks, incremented

File: tensorflow/inference/docker/2.16/py3/Dockerfile.cpu (hunk 1 above)
Comment (shantanutrip) [reply tensorflow/inference/docker/2.16/py3/Dockerfile.cpu]: Thanks, incremented

**Comment by shantanutrip** [comment]
> you can add the images to release_images_patches.yml in this PR as well

Thanks, added

**Review by junpuf — APPROVED** [review]

**Comment by arjkesh** [comment]
You can revert TOML with `python src/prepare_dlc_dev_environment.py -rcp origin`, assuming that your remote is called 'origin'

**Comment by shantanutrip** [comment]
> You can revert TOML with `python src/prepare_dlc_dev_environment.py -rcp origin`, assuming that your remote is called 'origin'

Nice feature! Thanks - I used it to revert temp changes.

**Review by arjkesh — APPROVED** [review]

**Review by junpuf — APPROVED** [review]

**Pull request merged**